dsl/flatten.cc

```diff
@@ -0,0 +1,304 @@
+// MethodDef) then we know that there are no MethodDefs lurking deeper in the tree. In order to work correctly, this
+// also needs to move some non-method-def things as well, specifically `sig`s and sends for method visibility
+// (e.g. `private` and the like), and it also updates the static-ness of some MethodDefs based on where they have
+// appeared in a nested context.
+//
+// So, a file like the following
+//
+// class A
+//   sig{void}
+//   private def foo
+//     sig{void}
+//     def self.bar; end
+//   end
+// end
+//
+// will morally be transformed into the following
+//
+// class A
+//   sig{void}
+//   private def foo; :bar; end
+//   sig{void}
+//   def bar; end   # notice the lack of `self.` here
+// end
+//
+// So no nested methods exist any longer, and additionally, the nested method `bar` has had the `self.` qualifier
+// removed: if you run the above code in Ruby, you'll find that `bar` is not defined as a class method on `A`, but
+// rather as a not-always-available instance method on `A`, so introducing it as a static method is not at all
+// correct. Finally, because methods evaluate to their corresponding symbols, the former location of `bar`'s definition
+// has been replaced with the verbatim symbol `:bar`.
+//
+// It does this by maintaining both a stack and a queue: elements on the stack include indices into the queue as well as
+// scoping information, while the queue consists of pointers to expressions; additionally, some stack elements do not
+// correspond to expressions that need to be moved, in which case their indices will be -1: for example, all methods at
+// the root of a class definition will have stack entries (so we can properly track nested methods) but none of them
+// will need to be moved, so they will have -1 indices.
+namespace sorbet::dsl {
+
+class FlattenWalk {
+    struct MethodData {
+        // This is an index into the methods stack
+        int idx;
+        // we need to keep information around about whether we're in a static outer context: for example, if we have
+        //
+        //   def self.foo; def bar; end; end
+        //
+        // then we should flatten it to
+        //
+        //  def self.foo; end
+        //  def self.bar; end
+        //
+        // which means when we get to `bar` we need to know that the outer context `foo` is static. We pass that down
+        // the current stack by means of this `isStatic` variable.
+        int staticLevel;
+        ast::Expression *target;
+        MethodData(int idx, int staticLevel, ast::Expression *target)
+            : idx(idx), staticLevel(staticLevel), target(target){};
+    };
+    struct MovedItem {
+        unique_ptr<ast::Expression> expr;
+        int staticLevel;
+        MovedItem(unique_ptr<ast::Expression> expr, int staticLevel) : expr(move(expr)), staticLevel(staticLevel){};
+        MovedItem() = default;
+    };
+    struct Methods {
+        vector<MovedItem> methods;
+        vector<MethodData> stack;
```

Comment:
what does this stack model?

Reply:
A single `Methods` instance corresponds to the body of a `ClassDef`; the `stack` corresponds to the nested methods inside of that `ClassDef`, and `methods` corresponds to the methods that we're moving out of nesting. If you never have any methods nested inside other methods, then that stack will never be more than one element deep. When we get to the point indicated by the comment here:
```ruby
class A
  class N; def foo; def bar; end; end; end
  class B
    def self.f
      def self.g
        # <- here
      end
    end
  end
end
```
we'll our overall state will look something like this:
```c++
[ Methods { // corresponding to A
    .methods = [ bar ], // this is the method we need to move to the end of
                        // A's body
    .stack = [],
  },
  Methods { // corresponding to B
    .methods = [], // at the point indicated above, we have not yet finished 
                   // processing #g but once we do we'll add it here
    .stack = [(false,1), (true,2)], // these correspond to #f and #g respectively
  }
```